DogStatsD Data aggregation (#5893)
DataDog/documentation: content/en/developers/faq/is-there-an-alternative-to-dogstatsd-and-the-api-to-submit-metrics-threadstats.md

2. You don't want to install the Datadog Agent
3. Or you want more flexibility than DogStatsD has to offer; for instance, you want to send metrics with timestamps

## Threadstats variations with DogStatsD

As in [DogStatsD](/developers/dogstatsd), Threadstats performs data aggregation for performance reasons.

### Variations

* The main difference is that metrics received by Threadstats may already have a timestamp
* Metrics are not aggregated via a centralized server, but they are aggregated and flushed in a Python thread of your script. So you'll get a per-script aggregation instead of a per host aggregation

To handle timestamps, Threadstats uses 2 parameters: a flush interval and a roll-up interval.

> * The flush interval defines the time interval between two consecutive {data aggregation + data submission}.

Reviewer: Unsure what is being said here. Is it between two consecutive data aggregation-submission instances?

Author: indeed, prior to be sent the data goes through 2 phases: 1. aggregation  then 2. submission. The flush interval is the sum of those two phases.

Reviewer: Let's change it to 
```suggestion
* The flush interval defines the time interval between two consecutive data aggregation-and-submission events.
```

* The roll up interval defines the data granularity after aggregation.

### Example with flush_interval=10 and roll_up_interval=5

> For instance, during the flush interval of 10 seconds (between 10:00:00 and 10:00:10), Threadstats receives 5 datapoints for the same metric name (a counter) and same tags, with {timestamps, values} being:

Reviewer: ```suggestion
Consider an instance where, during a flush interval of 10 seconds (between 10:00:00 and 10:00:10), Threadstats receives 5 datapoints for the same metric name (a counter) and same tags, with the following {`timestamp`, `value`} pairs:
```

> 1. {09:30:15, 1}, {10:00:00, 2}, {10:00:04,1}, {10:00:05,1}, {10:00:09,1} # 1- original datapoints

Reviewer: ```suggestion
`{09:30:15, 1}, {10:00:00, 2}, {10:00:04,1}, {10:00:05,1}, {10:00:09,1} `
```

> 2. {09:30:10, 1}, {10:00:00, 2}, {10:00:00,1}, {10:00:05,1}, {10:00:05,1} # 2- every datapoint in the same roll_up_interval (5 seconds) gets the same timestamp

Reviewer: ```suggestion
Next, every datapoint in the same `roll_up_interval` (5 seconds) gets the same timestamp:

`{09:30:10, 1}, {10:00:00, 2}, {10:00:00,1}, {10:00:05,1}, {10:00:05,1}`

```

> 3. {09:30:10, 1}, {10:00:00, 3}, {10:00:05,2} # 3- data is aggregated and only 4 values are eventually submitted to Datadog

Reviewer: ```suggestion
Finally, data is aggregated by timestamp. The following 4 values are then submitted to Datadog: 
`{09:30:10, 1}, {10:00:00, 3}, {10:00:05,2}`
```
